File: clients/js/websocket-client.js
const WebSocket = require('ws');

class MoonlightWebSocketClient {
    constructor(serverUrl, config) {
        this.serverUrl = serverUrl;
        this.config = config;
        this.ws = null;
        this.connected = false;
        this.reconnectInterval = 5000; // 5 seconds
        this.heartbeatInterval = 30000; // 30 seconds
        this.heartbeatTimer = null;
    }

    connect() {
        console.log(`Connecting to ${this.serverUrl}...`);

        this.ws = new WebSocket(this.serverUrl);

        this.ws.on('open', () => {
            console.log('WebSocket connection established');
            this.connected = true;

            // Register client
            this.register();

            // Start heartbeat
            this.startHeartbeat();
        });

        this.ws.on('message', (data) => {
            try {
                const message = JSON.parse(data);
                this.handleMessage(message);
            } catch (error) {
                console.error('Failed to parse message:', error);
            }
        });

        this.ws.on('close', () => {
            console.log('WebSocket connection closed');
            this.connected = false;
            this.stopHeartbeat();

            // Attempt to reconnect
            setTimeout(() => {
                this.connect();
            }, this.reconnectInterval);
        });

        this.ws.on('error', (error) => {
            console.error('WebSocket error:', error);
        });
    }

    register() {
        const registerMessage = {
            type: 'register',
            payload: {
                ip: this.config.ip,
                node_id: this.config.nodeId,
                token: this.config.token,
                region: this.config.region,
                port: this.config.port || 3000
            }
        };

        this.send(registerMessage);
    }

    startHeartbeat() {
        this.heartbeatTimer = setInterval(() => {
            if (this.connected) {
                const heartbeatMessage = {
                    type: 'heartbeat',
                    payload: {
                        timestamp: new Date().toISOString()
                    }
                };
                this.send(heartbeatMessage);
            }
        }, this.heartbeatInterval);
    }

    stopHeartbeat() {
        if (this.heartbeatTimer) {
            clearInterval(this.heartbeatTimer);
            this.heartbeatTimer = null;
        }
    }

    handleMessage(message) {
        switch (message.type) {
            case 'registered':
                console.log('Successfully registered with server');
                break;

            case 'heartbeat_ack':
                console.log('Heartbeat acknowledged');
                break;

            case 'task':
                this.handleTask(message);
                break;

            case 'error':
                console.error('Server error:', message.payload);
                break;

            default:
                console.log('Unknown message type:', message.type);
        }
    }

    async handleTask(message) {
        console.log(`Received task ${message.task_id}:`, message.payload);

        try {
            // Process the task (this is where your actual work happens)
            const result = await this.processTask(message.payload);

            // Send task response
            const responseMessage = {
                type: 'task_response',
                payload: {
                    task_id: message.task_id,
                    response: result.data,
                    status: result.status,
                    headers: result.headers || {}
                }
            };

            this.send(responseMessage);
            console.log(`Task ${message.task_id} completed successfully`);

        } catch (error) {
            console.error(`Task ${message.task_id} failed:`, error);

            // Send error response
            const errorResponse = {
                type: 'task_response',
                payload: {
                    task_id: message.task_id,
                    response: { error: error.message },
                    status: 500,
                    headers: { 'Content-Type': 'application/json' }
                }
            };

            this.send(errorResponse);
        }
    }

    async processTask(payload) {
        // This is where you implement your actual task processing logic
        // For this example, we'll just echo back the payload
        console.log("Type: ", typeof payload);
        console.log("Payload: ", payload);

        // Simulate some processing time
        await new Promise(resolve => setTimeout(resolve, 100));

        // Return raw data instead of pre-serialized JSON to avoid double serialization
        return {
            data: {
                processed: true,
                timestamp: new Date().toISOString(),
                data: payload
            },
            status: 200,
            headers: { 'Content-Type': 'application/json' }
        };
    }

    send(message) {
        if (this.connected && this.ws) {
            this.ws.send(JSON.stringify(message));
        } else {
            console.error('Cannot send message: not connected');
        }
    }

    disconnect() {
        this.stopHeartbeat();
        if (this.ws) {
            this.ws.close();
        }
    }
}

// Example usage
if (require.main === module) {
    const config = {
        ip: '192.168.1.100',
        nodeId: 'node-002',
        token: 'supersecrettokenox0',
        region: 'us-west',
        port: 3000
    };

    const client = new MoonlightWebSocketClient('ws://localhost:8080/ws', config);
    client.connect();

    // Handle graceful shutdown
    process.on('SIGINT', () => {
        console.log('Shutting down...');
        client.disconnect();
        process.exit(0);
    });
}

module.exports = MoonlightWebSocketClient;

File: clients/python/websocket-client.py
#!/usr/bin/env python3
"""
Moonlight WebSocket Client for Python
"""

import asyncio
import json
import logging
import time
from typing import Dict, Any
import websockets

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


class MoonlightWebSocketClient:
    def __init__(self, server_url: str, config: Dict[str, Any]):
        self.server_url = server_url
        self.config = config
        self.websocket = None
        self.connected = False
        self.reconnect_interval = 5  # seconds
        self.heartbeat_interval = 30  # seconds
        self.heartbeat_task = None

    async def connect(self):
        """Connect to the WebSocket server"""
        while True:
            try:
                logger.info(f"Connecting to {self.server_url}...")
                self.websocket = await websockets.connect(self.server_url)
                self.connected = True
                logger.info("WebSocket connection established")

                # Register client
                await self.register()

                # Start heartbeat
                self.heartbeat_task = asyncio.create_task(self.heartbeat_loop())

                # Listen for messages
                await self.message_loop()

            except websockets.exceptions.ConnectionClosed:
                logger.info("WebSocket connection closed")
                self.connected = False
                if self.heartbeat_task:
                    self.heartbeat_task.cancel()

            except Exception as e:
                logger.error(f"Connection error: {e}")

            finally:
                self.connected = False
                if self.websocket:
                    await self.websocket.close()

            # Wait before reconnecting
            logger.info(f"Reconnecting in {self.reconnect_interval} seconds...")
            await asyncio.sleep(self.reconnect_interval)

    async def register(self):
        """Register client with the server"""
        register_message = {
            "type": "register",
            "payload": {
                "ip": self.config["ip"],
                "node_id": self.config["node_id"],
                "token": self.config["token"],
                "region": self.config["region"],
                "port": self.config.get("port", 3000)
            }
        }
        await self.send(register_message)

    async def heartbeat_loop(self):
        """Send periodic heartbeat messages"""
        while self.connected:
            try:
                heartbeat_message = {
                    "type": "heartbeat",
                    "payload": {
                        "timestamp": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime())
                    }
                }
                await self.send(heartbeat_message)
                await asyncio.sleep(self.heartbeat_interval)
            except Exception as e:
                logger.error(f"Heartbeat error: {e}")
                break

    async def message_loop(self):
        """Listen for incoming messages"""
        try:
            async for message in self.websocket:
                try:
                    data = json.loads(message)
                    await self.handle_message(data)
                except json.JSONDecodeError as e:
                    logger.error(f"Failed to parse message: {e}")
        except Exception as e:
            logger.error(f"Message loop error: {e}")

    async def handle_message(self, message: Dict[str, Any]):
        """Handle incoming messages"""
        msg_type = message.get("type")
        
        if msg_type == "registered":
            logger.info("Successfully registered with server")
            
        elif msg_type == "heartbeat_ack":
            logger.info("Heartbeat acknowledged")
            
        elif msg_type == "task":
            await self.handle_task(message)
            
        elif msg_type == "error":
            logger.error(f"Server error: {message.get('payload')}")
            
        else:
            logger.warning(f"Unknown message type: {msg_type}")

    async def handle_task(self, message: Dict[str, Any]):
        """Handle incoming task"""
        task_id = message.get("task_id")
        payload = message.get("payload")
        
        logger.info(f"Received task {task_id}: {payload}")
        
        try:
            # Process the task
            result = await self.process_task(payload)
            
            # Send task response
            response_message = {
                "type": "task_response",
                "payload": {
                    "task_id": task_id,
                    "response": result["data"],
                    "status": result["status"],
                    "headers": result.get("headers", {})
                }
            }
            
            await self.send(response_message)
            logger.info(f"Task {task_id} completed successfully")
            
        except Exception as e:
            logger.error(f"Task {task_id} failed: {e}")
            
            # Send error response
            error_response = {
                "type": "task_response",
                "payload": {
                    "task_id": task_id,
                    "response": json.dumps({"error": str(e)}),
                    "status": 500,
                    "headers": {"Content-Type": "application/json"}
                }
            }
            
            await self.send(error_response)

    async def process_task(self, payload: str) -> Dict[str, Any]:
        """Process a task (implement your actual logic here)"""
        # Parse the payload
        
        # Simulate some processing time
        await asyncio.sleep(0.1)
        
        # For this example, we'll just echo back the payload
        return {
            "data": json.dumps({
                "processed": True,
                "original": payload,
                "timestamp": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime())
            }),
            "status": 200,
            "headers": {"Content-Type": "application/json"}
        }

    async def send(self, message: Dict[str, Any]):
        """Send a message to the server"""
        if self.connected and self.websocket:
            await self.websocket.send(json.dumps(message))
        else:
            logger.error("Cannot send message: not connected")

    async def disconnect(self):
        """Disconnect from the server"""
        self.connected = False
        if self.heartbeat_task:
            self.heartbeat_task.cancel()
        if self.websocket:
            await self.websocket.close()


async def main():
    """Main function"""
    config = {
        "ip": "192.168.1.100",
        "node_id": "node-003",
        "token": "supersecrettokenox1",
        "region": "us-west",
        "port": 3000
    }
    
    client = MoonlightWebSocketClient("ws://localhost:8080/ws", config)
    
    try:
        await client.connect()
    except KeyboardInterrupt:
        logger.info("Shutting down...")
        await client.disconnect()


if __name__ == "__main__":
    asyncio.run(main())

